internal/manifests/mutate.go

```diff
@@ -0,0 +1,85 @@
+package manifests
+
+import (
+	"fmt"
+	"reflect"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// MutateFuncFor returns a mutate function based on the
+// existing resource's concrete type. It supports currently
+// only the following types or else panics:
+// - ConfigMap
+// - Service
+// - Deployment
+// - StatefulSet
+func MutateFuncFor(existing, desired client.Object) controllerutil.MutateFn {
+	return func() error {
+		existing.SetAnnotations(desired.GetAnnotations())
+		existing.SetLabels(desired.GetLabels())
+
+		switch existing.(type) {
+		case *corev1.ConfigMap:
+			cm := existing.(*corev1.ConfigMap)
+			wantCm := desired.(*corev1.ConfigMap)
+			mutateConfigMap(cm, wantCm)
+
+		case *corev1.Service:
+			svc := existing.(*corev1.Service)
+			wantSvc := desired.(*corev1.Service)
+			mutateService(svc, wantSvc)
+
+		case *appsv1.Deployment:
+			dpl := existing.(*appsv1.Deployment)
+			wantDpl := desired.(*appsv1.Deployment)
+			mutateDeployment(dpl, wantDpl)
+
+		case *appsv1.StatefulSet:
+			sts := existing.(*appsv1.StatefulSet)
+			wantSts := desired.(*appsv1.StatefulSet)
+			mutateStatefulSet(sts, wantSts)
+
+		default:
+			t := reflect.TypeOf(existing).String()
+			err := fmt.Errorf("missing mutate implementation for resource type %s", t) //nolint:goerr113
+			panic(err)
```

Comment:
You can't panic. We have to figure something else out here. Change the signature of MuteFuncFor to `controllerutil.MutateFn, error`

Reply:
https://golang.org/doc/effective_go#panic

This is not a compile-time issue that would always be caught. The tests capture the problem now, but panics _could_ happen in a production environment and this destroys any traceability we can gain from error handling such as logs, etc. Panics explicitly exit the application and bypass any upstream handling like closing clients, etc.